Feat/boolean ops

rmanoka opened the pull request:
- [x] I agree to follow the project's [code of conduct](https://github.com/georust/geo/blob/main/CODE_OF_CONDUCT.md).
- [x] I added an entry to `CHANGES.md` if knowledge of this change could be valuable to users.
---

PR to import the boolean-ops code from [geo-crossings].  The impl. is not a 100% accurate to fixed-precision errors, but seems to handle the degenerate cases encountered in [rust-geo-booleanop].  When comparing the difference in outputs between the two algos, both algos panic in a few cases.  The hope is to eventually improve some of the internal data-structures to be a 100% robust. 

The performance, is about 1.25-1.5x worse when using the splay-tree, and about 1.5-2x if using vanilla btree.  However, splay-tree seems to trigger an extra panic (not on the test cases, but while comparing outputs between the two impls), so I'm inclined to stick with vanilla BTree.  Eventually, I hope we can figure out a intrusive impl. of BTree (eg. `btree_slab`) that is both performant, and fixes the fp accuracy issues.

[rust-geo-booleanop]: https://github.com/21re/rust-geo-booleanop
[geo-crossings]: https://github.com/rmanoka/geo-crossings

rmanoka commented:
First cut port of the code is done: the tests all pass, and the benches are about the same as in geo-crossings.  There's a bit of a code-bloat especially on the test side as we import a bunch of extra crates to compare outputs here vs the `rust-geo-booleanop` reference impl.  I will work on streamlining them and listing key / controversial changes in the next few weeks.

michaelkirk commented on geo/src/algorithm/sweep/line_or_point.rs:
```diff
@@ -0,0 +1,259 @@
+                if p1 > p2 {
+                    return other.partial_cmp(self).map(Ordering::reverse);
+                }
+                if p1 >= q2 || p2 >= q2 {
```
> p2 >= q2

Isn't `p2 >= q2` false by construction?

rmanoka replied:
Nice catch,  It should be `p2 >= q1`

michaelkirk reviewed (commented):
I'm still mid review, but stopping for the day. I need to actually read the paper before diving into the substance of the algorithm. 😅 

Excited to see the PR!

rmanoka commented on geo/Cargo.toml:
```diff
@@ -32,6 +33,18 @@ geo-test-fixtures = { path = "../geo-test-fixtures" }
+serde_json = "1.0.81"
+serde_derive = "1.0.137"
+glob = "0.3.0"
+gt_prev = { version = "0.6.2" , package = "geo-types" }
```
Currently, we use rust-geo-booleanop for comparison tests, and for benchmarks.  That crate only supports the older version of geo-types.  Should we move the benches to a separate crate?  The tests should not need to call the other impl. really.

rmanoka replied:
Moved the tests and benches to a separate crate, so all these deps have moved there.  We're still adding `wkt` as a dev-dep as it's useful to capture test-case geoms succinctly.

rmanoka commented on geo/benches/utils/random.rs:
```diff
@@ -0,0 +1,117 @@
+// random-geom-samplers as `Distribution::sample` implementations directly. But
+// that needs to happen in geo-types. May be related to our discussion on
+// handling private-utils in geo-types, but this could be folded into a
+// feature-flag too.
```
Ah, I didn't realize the connection.  I don't understand `Arbitrary` well, but with random-sampling, we want points/geoms with guarantees: eg. that it falls within a given bounding box, etc;  it seems like the former is about converting valid random-bits to a data-type, so I don't know how to use it.  Will try experimenting with the traits and see if they fit well in this repo.

rmanoka replied:
Handling this in a future PR; it's a separate concern, and not a blocker here.

rmanoka commented on geo/src/algorithm/sweep/point.rs:
```diff
@@ -0,0 +1,90 @@
+///
+/// Note that the scalar type `T` is only required to implement `PartialOrd`.
+/// Thus, it is a logical error to construct this struct unless the coords are
+/// guaranteed to be orderable.
```
Yes, it's not enforced but expected.

rmanoka commented on geo/src/algorithm/sweep/line_or_point.rs:
```diff
@@ -0,0 +1,259 @@
+        let line = other.line();
+        if !self.is_line() {
+            let p = self.left;
+            if <T as HasKernel>::Ker::orient2d(line.start, *p, line.end) == Orientation::Collinear {
```
The conversions should get optimized to no-ops right?

michaelkirk replied:
I'm not sure which conversions you're asking about.

To clarify, my take-it-or-leave-it suggestion was to replace the above with something like:

```rust
let line = other.line();
if !self.is_line() {
    let p = self.left;
    if line.intersects(&*p) {
        Some(*self)
    }  else {
        None
    }
} else {
    line_intersection(self.line(), line).map(|l| match l {
        LineIntersection::SinglePoint { intersection, .. } => intersection.into(),
        LineIntersection::Collinear { intersection } => intersection.into(),
    })
}
```

Which doesn't introduce any new conversions - I just liked it because it's more succinct.

rmanoka replied:
Makes sense; done!

rmanoka commented on geo/src/algorithm/sweep/line_or_point.rs:
```diff
@@ -0,0 +1,259 @@
+
+    #[cfg(test)]
+    pub fn coords_equal(&self, other: &LineOrPoint<T>) -> bool {
+        self.is_line() == other.is_line() && self.end_points() == other.end_points()
```
Sorry, `end_points` is as used in math (so both vertices of the line are "end points").  It's only used in tests, because the `PartialEq` has to agree with `PartialOrd` and accepts more segments as being "equal" (anything that overlaps)

michaelkirk replied:
oh right - end_point**s** I read it as `end_point`! My mistake.

rmanoka commented on geo/src/algorithm/sweep/line_or_point.rs:
```diff
@@ -0,0 +1,259 @@
+                    Some(Ordering::Equal)
+                } else {
+                    // Unequal points do not satisfy pre-condition and
+                    // can't be ordered.
```
This is `PartialOrd` so `None` is a valid answer.  We anyway debug_assert elsewhere where we expect them to be totally-ordered.

rmanoka commented on geo/src/algorithm/sweep/line_or_point.rs:
```diff
@@ -0,0 +1,259 @@
+        self.left != self.right
+    }
+
+    /// Return a [`Line`] if it is one, otherwise `None`.
```
Good catch; the docs needed adjustment.

rmanoka commented:
Upgrading to rstar 0.9.x causes the `concave_hull_norway_test` test to fail!  Should I upgrade the output fixture?  It may be bug-fixes in rstar, but I'm not sure.

michaelkirk commented:
> Upgrading to rstar 0.9.x causes the concave_hull_norway_test test to fail! Should I upgrade the output fixture? It may be bug-fixes in rstar, but I'm not sure.

heh, it's always more complicated than you anticipate. 😄 

I assume we'll eventually want to update rstar, but it'd be good to investigate why the change in the test case. In order not to block you I'd be fine with either *not* updating rstar yet (and copying in the `GeomWithData` as you originally had it)

*or* spending some time convincing ourselves that the difference is OK.

I'll have some time tomorrow that I could look into this if you'd like.

rmanoka commented:
> > Upgrading to rstar 0.9.x causes the concave_hull_norway_test test to fail! Should I upgrade the output fixture? It may be bug-fixes in rstar, but I'm not sure.
> 
> heh, it's always more complicated than you anticipate. 😄
:)

> I'll have some time tomorrow that I could look into this if you'd like to delegate.

Sounds good, I'll give it a couple of days and revert to previous rstar o/w.

michaelkirk commented on geo/src/algorithm/bool_ops/op.rs:
```diff
@@ -0,0 +1,281 @@
+#[derive(Clone)]
+struct Edge<T: Float> {
+    geom: LineOrPoint<T>,
+    is_first: bool,
```
I confess that I'm confused about the distinction between  edge._region.is_first vs edge.is_first - can you elucidate?

rmanoka replied:
An edge belongs to the first-input geometry or the second, so we use a flag.  A region is space between edges, and can belong to either/both/none of the geometries: 4 combinations -> two flags within Region.

michaelkirk reviewed (commented):
I'm still making my way through understanding this. I'm sorry it's taking me so long, but it's pretty dense and seems to be doing more than what I expected from just reading the original paper. 

Here's some more feedback so far.

rmanoka commented on geo/src/algorithm/bool_ops/op.rs:
```diff
@@ -0,0 +1,281 @@
+    }
+    fn add_closed_ring(&mut self, ring: &LineString<T>, is_first: bool, _is_hole: bool) {
+        assert!(ring.is_closed());
+        if ring.coords_count() <= 3 {
```
Rings are closed, so the first and last are the same.

rmanoka commented on geo/src/algorithm/sweep/iter.rs:
```diff
@@ -0,0 +1,409 @@
+            max_size.unwrap_or(min_size)
+        };
+        let sweep = Sweep::new(iter);
+        let segments = Vec::with_capacity(4 * size);
```
:) Not really.  I didn't benchmark it, as this allocation happens only once per run.  I think 2*size is the max upper-bound on the size of the segments during any run, but wanted to be safer.

rmanoka commented on geo/src/algorithm/bool_ops/op.rs:
```diff
@@ -0,0 +1,281 @@
+        // the segment. Thus, we detect this and store both sides of the region.
+        // Finally, note that we need to store both sides of the segment, as
+        // this cannot be computed from the current edge alone (it may depend on
+        // more overlapping edges).
```
The impl. here is primary based on Martinez-Rueda-Freito; which is an extension of Martinez-Rueda; I did re-work many of the impl. details to convince myself.  I did not find much resources on handling fixed-precision errors, so most of that is our own "research" (a.k.a debugging).

rmanoka commented on geo/src/algorithm/sweep/cross.rs:
```diff
@@ -0,0 +1,66 @@
+/// also impl. `Clone`. If the custom type is not cheap to clone, use
+/// either a reference to the type, a [`Rc`] or an [`Arc`]. All these
+/// are supported via blanket trait implementations.
+pub trait Cross: Sized + Debug {
```
Yeah, the earlier version called this trait `Crossable` but it was a mouthful and I decided for a shorter-name as opposed to a grammatically appropriate one.  Will fix the docs.

rmanoka commented on geo/src/algorithm/bool_ops/op.rs:
```diff
@@ -0,0 +1,281 @@
+        }
+    }
+
+    pub(crate) fn add_multi_polygon(&mut self, mp: &MultiPolygon<T>, is_first: bool) {
```
Yes, `is_first` here refers to which input it comes from.

rmanoka commented on geo/src/algorithm/bool_ops/op.rs:
```diff
@@ -0,0 +1,281 @@
+                    }
+                })
+            }
+            iter.intersections_mut().sort_unstable_by(compare_crossings);
```
Yeah, their impls. use a more intrusive sweep algo: the ordering of the points is not based only on the end-point, but based on the ordering of the segment.  In our impl., the sweep was coded earlier, and I wanted to re-use it.  

I think the asymptotics don't change: say the total K intersections is returned as packets over T points: k_1...k_T (they sum to K).  Then, we can show that $ \sum { k_i log(k_i) }$ is maximized when all k_i are equal, and the complexity is still O(K log K).

I like keeping the sweep and the rest separate, at least until we can iron out all precision issues.

rmanoka commented on .github/workflows/test.yml:
```diff
@@ -67,9 +67,8 @@ jobs:
           # Minimum supported rust version (MSRV)
-          - "georust/geo-ci:proj-9.0.0-rust-1.56"
-          # Two most recent releases - we omit older ones for expedient CI
           - "georust/geo-ci:proj-9.0.0-rust-1.58"
```
Done.

rmanoka commented on geo/src/algorithm/sweep/active.rs:
```diff
@@ -0,0 +1,98 @@
+impl<T: PartialOrd> Ord for Active<T> {
+    fn cmp(&self, other: &Self) -> Ordering {
+        T::partial_cmp(self, other).unwrap()
+        // self.partial_cmp(other).unwrap()
```
Done

michaelkirk commented:
I pushed up a branch which merges with master and integrates some of the JTS Overlay test cases here: a65e9078

Some of them are failing. I haven't looked into most of them to determine if they reflect "poor" performance on our part or just "different assumptions". I personally don't necessarily consider those failures to be a blocker.  If I'm understanding, the current implementation you have here is already passing the test suite of the current "go to" solution for boolean ops in rust - and even passing a few additional tests.

For that reason I don't think the failing JTS tests should block merging this - but I'd kind of like to incorporate them somehow. At least the passing subset. WDYT?

michaelkirk commented:
> I'd kind of like to incorporate the JTS tests somehow. At least the passing subset. WDYT?

To be clear - I'd be happy to do this in a follow up PR.

michaelkirk reviewed (approved):
I have no blocking concerns about this, and am in fact quite excited to get boolean operations in geo! 

It's been sitting here a while, so I'm sure we're both eager to see it merged. Did you want anyone else to review it? It's a a pretty important feature with a lot of code so it'd be nice to have more eyes on it, but at the same time, there's decent test coverage and the proposed API seems really clean and minimal.

rmanoka commented:
Thanks @michaelkirk ;  I've also been wanting to test on the JTS test-suites.  Note that the output may not be exactly the same, even upto rotation of the vertices.  It's best to compute the difference and check the relative area of the difference to the expected is less than a threshold.  That might help with at least a few failures (we do that with the rust-geo-booleanop test cases too).

michaelkirk commented:
I see how that could be a good way to test that the results are "approximately correct". 

Once concern would be that, even if the magnitude of the error were small, it might affect topology. It'd be desirable to be topologically tight wherever possible, so I wouldn't necessarily want to sweep topological diversions under the rug, just because the `Area` of the diversion is small.

rmanoka commented:
Yes, that's true;  I also realize we're using relative_eq in the rotated-polygon-check, so it should already work at least with slightly larger epsilon, max_relative params.

michaelkirk commented:
BTW - I looked into the rstar 0.8 vs 0.9 discrepancy.

I didn't look into the root cause of the difference, but I do feel that the actual variance is not problematic, and that we should update to rstar-9, and just update the expected output. The output still appears to be a proper hull, and the "area" covered by the discrepancy is pretty small:  0.00008449 the size of the original output.

Here's a series of "zoom ins" for context in the affected area (there's only one region)

<img width="707" alt="Screen Shot 2022-06-08 at 1 19 13 AM" src="https://user-images.githubusercontent.com/217057/172537780-0891d38a-81e6-4b1d-9b3a-492d1e041338.png">
<img width="857" alt="Screen Shot 2022-06-08 at 1 19 23 AM" src="https://user-images.githubusercontent.com/217057/172537779-ce02c9f1-722b-4cc5-add6-1891287d1f73.png">
<img width="864" alt="Screen Shot 2022-06-08 at 1 19 39 AM" src="https://user-images.githubusercontent.com/217057/172537777-5a0c1c0c-82d0-4fde-b6c1-f51513fe369a.png">
<img width="761" alt="Screen Shot 2022-06-08 at 1 19 50 AM" src="https://user-images.githubusercontent.com/217057/172537776-2736240f-25be-4833-bdfc-ffbf4f659ba1.png">


Here's the code I used to generate the magnitude of the discrepancy:
```
use geo::{Geometry, Area, Transform};

fn main() {
    fn area(file: &str) -> f64{
        let geojson_str = std::fs::read_to_string(file).unwrap();
        use std::str::FromStr;
        let geojson = geojson::GeoJson::from_str(&geojson_str).unwrap();
        let poly = Geometry::<f64>::try_from(geojson).unwrap();
        let proj_poly = poly.transformed_crs_to_crs("WGS84", "ESRI:102101").unwrap();
        println!("area: {}, proj_area: {}", poly.unsigned_area(), proj_poly.unsigned_area());
        proj_poly.unsigned_area()
    }

    let diff = area("symdiff.geojson");
    let total = area("concave-hull-rstar-0.9.geojson");
    println!("{} / {} = {:.8}", diff, total, diff / total);
}
```

input if you want to reproduce it: [Archive.zip](https://github.com/georust/geo/files/8858522/Archive.zip)


And the output:
> area: 0.00804129116421399, proj_area: 52821317.77513506
> area: 136.57700656512267, proj_area: 625147693604.5656
> 52821317.77513506 / 625147693604.5656 = 0.00008449

rmanoka commented on geo/src/algorithm/bool_ops/rings.rs:
```diff
@@ -0,0 +1,259 @@
+
+use crate::sweep::{LineOrPoint, SweepPoint};
+
+#[derive(Debug, Clone)]
+pub struct Rings<T: GeoNum> {
+    chains: Vec<Chain<T>>,
+    end_points: BTreeMap<SweepPoint<T>, Link>,
+}
+
+impl<T: GeoNum> Default for Rings<T> {
+    fn default() -> Self {
+        Self {
+            chains: Default::default(),
+            end_points: Default::default(),
+        }
+    }
+}
```
:) yes.

rmanoka replied:
Oh right, the derive macros don't infer the correct trait bounds.  Here, it adds an extra `T: Default` which we want to avoid.

rmanoka commented:
Have added the doc suggestions, and adjusted the concave-hull for rstar update.  Also merged and added fixes for changes from main (rebase gave conflicts, and I wanted to be safe with code changes).

I've also made some partial progress on using a custom-rolled b-tree so we've a framework to address the fp-errors [here](https://github.com/rmanoka/geo/tree/misc/sweep-tree) but it's quite bleeding edge, so I don't plan to address it pull it in here.

Reg. the JTS tests, I'm unsure if I can get to it quickly, so would suggest punting to another PR.

rmanoka commented:
I'll give it a day and merge this, if there are no further suggestions.  @frewsxcv @urschrei @lnicola

kylebarron commented on rfcs/2022-05-24-boolean-ops.md:
```diff
@@ -0,0 +1,18 @@
+- Feature Name: `boolean-ops`
+- Start Date: 2022-05-24
+- Feature PR: //github.com/georust/geo/pull/835
```
Just a note: the below links seem to be rendered correctly, but github's markdown renderer doesn't like this link: 
<img width="304" alt="image" src="https://user-images.githubusercontent.com/15164633/173126559-43728a3c-8e11-4e07-9b41-5a4348e09756.png">

rmanoka commented:
bors r=michaelkirk
